File: frontend/src/models/BagItem.ts
export interface BagItem {
  characterName: string; 
  name?: string;
  description?: string;
  id: number;
  icon: string;
  count: number;
  charges?: number;
  infusions?: number[];
  upgrades?: number[];
  skin?: number;
  stats?: { [key: string]: unknown };
  dyes?: number[];
  binding?: string;
  boundTo?: string;
  rarity?: string;
}

export interface APIBagItem {
  character_name: string; 
  name?: string;
  description?: string;
  id: number;
  icon: string;
  count: number;
  charges?: number;
  infusions?: number[];
  upgrades?: number[];
  skin?: number;
  stats?: { [key: string]: unknown };
  dyes?: number[];
  binding?: string;
  bound_to?: string;
  rarity?: string;
}

export function APIBagItemToBagItem(apiBagItem: APIBagItem): BagItem {
return {
  characterName: apiBagItem.character_name, 
  name: apiBagItem.name,
  description: apiBagItem.description,
  id: apiBagItem.id,
  icon: apiBagItem.icon,
  count: apiBagItem.count,
  charges: apiBagItem.charges,
  infusions: apiBagItem.infusions,
  upgrades: apiBagItem.upgrades,
  skin: apiBagItem.skin,
  stats: apiBagItem.stats,
  dyes: apiBagItem.dyes,
  binding: apiBagItem.binding,
  boundTo: apiBagItem.bound_to,
  rarity: apiBagItem.rarity as string,
}

}

File: backend/internal/api/models/bagItem.go
package models

import (
	"database/sql/driver"
	"encoding/json"
	"errors"
	"fmt"
)

type BagItem struct {
	CharacterName string                  `json:"character_name"`
	Name          *string                 `json:"name"`
	Description   *string                 `json:"description"`
	BagItemID     uint                    `json:"id"`
	Icon          string                  `json:"icon"`
	Count         uint                    `json:"count"`
	Charges       *uint                   `json:"charges,omitempty"`
	Infusions     *[]int64                `json:"infusions,omitempty"`
	Upgrades      *[]int64                `json:"upgrades,omitempty"`
	Skin          *uint                   `json:"skin,omitempty"`
	Stats         *map[string]interface{} `json:"stats,omitempty" gorm:"type:json"`
	Dyes          *[]int64                `json:"dyes,omitempty" gorm:"type:integer[]"`
	Binding       *string                 `json:"binding,omitempty"`
	BoundTo       *string                 `json:"bound_to,omitempty"`
	Rarity        *string                 `json:"rarity"`
	Slot          *string                 `json:"slot"`
	Location      *string                 `json:"location"`
}

func (item BagItem) IsEquipment() bool {
	if item.Slot != nil && *item.Slot != "" {
		return true
	} else {
		return false
	}
}

type DetailsMap map[string]interface{}

func (detailsMap *DetailsMap) Scan(value interface{}) error {
	if value == nil {
		*detailsMap = nil
		return nil
	}
	bytes, ok := value.([]byte)
	if !ok {
		return errors.New(fmt.Sprint("Failed to unmarshal JSONB value:", value))
	}
	err := json.Unmarshal(bytes, detailsMap)
	return err

}

func (detailsMap DetailsMap) Value() (driver.Value, error) {
	if len(detailsMap) == 0 {
		return nil, nil
	}
	return json.Marshal(detailsMap)
}

type DetailsMapArray []map[string]interface{}

func (detailsMapArray *DetailsMapArray) Scan(value interface{}) error {
	if value == nil {
		*detailsMapArray = nil
		return nil
	}
	bytes, ok := value.([]byte)
	if !ok {
		return errors.New(fmt.Sprint("Failed to unmarshal JSONB value:", value))
	}
	err := json.Unmarshal(bytes, detailsMapArray)
	return err

}

func (detailsMapArray DetailsMapArray) Value() (driver.Value, error) {
	if len(detailsMapArray) == 0 {
		return nil, nil
	}
	return json.Marshal(detailsMapArray)
}
